feat(core): add support for fallback content in ng-content

crisbeto opened the pull request:
Adds the ability to specify content that Angular should fall back to if nothing is projected into an `ng-content` slot. For example, if we have the following setup

```
@Component({
  selector: 'my-comp',
  template: `
    <ng-content select="header">Default header</ng-content>
    <ng-content select="footer">Default footer</ng-content>
  `
})
class MyComp {}

@Component({
  template: `
    <my-comp>
      <footer>New footer</footer>
    </my-comp>
  `
})
class MyApp {}
```

The instance of `my-comp` in the app will have the default header and the new footer.

**Note:** Angular's content projection happens during creation time. This means that dynamically changing the contents of the slot will not cause the default content to show up, e.g. if a `if` block goes from `true` to `false`.

Fixes #12530.

crisbeto committed:
refactor(compiler-cli): capture ng-content children in AST

Updates the template AST to capture the content of `ng-content` elements instead of throwing an error.

crisbeto committed:
refactor(compiler): pass default content to projection instruction

Updates the code that generates the `projection` instruction to pass the template function containing the default content into it.

crisbeto committed:
refactor(compiler-cli): integrate fallback content for ng-content into template type checker

Adds logic to ingest the content of an `ng-content` element in the template type checker. We treat `ng-content` as a `ScopedNode`, because its content is inserted conditionally.

crisbeto committed:
feat(core): add support for fallback content in ng-content

Adds the ability to specify content that Angular should fall back to if nothing is projected into an `ng-content` slot. For example, if we have the following setup

```
@Component({
  selector: 'my-comp',
  template: `
    <ng-content select="header">Default header</ng-content>
    <ng-content select="footer">Default footer</ng-content>
  `
})
class MyComp {}

@Component({
  template: `
    <my-comp>
      <footer>New footer</footer>
    </my-comp>
  `
})
class MyApp {}
```

The instance of `my-comp` in the app will have the default header and the new footer.

**Note:** Angular's content projection happens during creation time. This means that dynamically changing the contents of the slot will not cause the default content to show up, e.g. if a `if` block goes from `true` to `false`.

Fixes #12530.

crisbeto commented on packages/core/test/acceptance/content_spec.ts:
```diff
@@ -384,12 +384,12 @@ describe('projection', () => {
       selector: 'child',
       template:
-          `<div *ngFor="let item of [1, 2]; let i = index">({{i}}):<ng-content></ng-content></div>`
+          '<div *ngFor="let item of [1, 2]; let i = index">({{i}}):<ng-content></ng-content></div>'
```
This change is unrelated, it was just messing up the syntax highlighting for the rest of the file for some reason.

pkozlowski-opensource commented on packages/core/src/render3/instructions/projection.ts:
```diff
@@ -109,31 +115,66 @@ export function ɵɵprojectionDef(projectionSlots?: ProjectionSlots): void {
   const hydrationInfo = lView[HYDRATION];
   const isNodeCreationMode = !hydrationInfo || isInSkipHydrationBlock();
-  if (isNodeCreationMode &&
+  const componentNode = lView[DECLARATION_COMPONENT_VIEW][T_HOST] as TElementNode;
```
super nit: how about `componentHostNode` or sth else that indicates a host?

pkozlowski-opensource commented on packages/core/test/acceptance/content_spec.ts:
```diff
@@ -1368,4 +1368,400 @@ describe('projection', () => {
+              `Catch-all fallback|Two fallback<div three="">Three</div></projection>`);
+    });
+
+    it('should render the catch-all slots fallback content if the element only contains comments',
```
Great test, thank you for thinking about this one!

pkozlowski-opensource commented on packages/core/test/acceptance/content_spec.ts:
```diff
@@ -1368,4 +1368,400 @@ describe('projection', () => {
+      fixture.detectChanges();
+      expect(getElementHtml(fixture.nativeElement)).toContain(`<projection>Fallback</projection>`);
+    });
+  });
```
Excellent tests @crisbeto! The only other cases I could think of are around re-projection. We should add a test verifying proper behaviour of the projection node with default content being re-projected (with the default content I presume?)

pkozlowski-opensource reviewed (approved):
LGTM overall, thank you for adding very thorough tests. The only addition I could think of are tests for re-projection with the default content.

crisbeto committed:
refactor(core): move template declaration logic into reusable function

Moves the logic that declares a template out into a separate function so that it can be reused in other places like control flow and defer, without having to call directly into the `template` instruction.

crisbeto committed:
fix(core): account for re-projected ng-content elements with fallback content

Passes the attributes of the `ng-content` element to the container that is created for the fallback content so that it can be re-projected into the same slot.

atscott reviewed (approved):
reviewed-for: language-service

dylhunn commented:
This PR was merged into the repository by commit 2fc11eae9ea65160866bf7ba46c10520ae9a141f.
